### PR TITLE
fix: ass subtitle render wrong when play half-sbs/half-ou 3d movie

### PR DESCRIPTION
**Pull request by taxigps**
## Description

see https://github.com/xbmc/xbmc/issues/14789
## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp [hunk 1]
```diff
@@ -343,6 +345,13 @@ COverlay* CRenderer::Convert(CDVDOverlaySSA* o, double pts)
   int targetWidth = m_rv.Width();
   int targetHeight = m_rv.Height();
   int useMargin;
+  int videomode = CServiceBroker::GetGUI()->GetStereoscopicsManager().GetStereoModeOfPlayingVideo();
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: Although `RENDER_STEREO_MODE` is an enum which can be converted to int, I would use the correct type.
```suggestion
  RENDER_STEREO_MODE videomode = CServiceBroker::GetGUI()->GetStereoscopicsManager().GetStereoModeOfPlayingVideo();
```

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp [hunk 2]
```diff
@@ -343,6 +345,13 @@ COverlay* CRenderer::Convert(CDVDOverlaySSA* o, double pts)
...
+  // Render subtitle of half-sbs and half-ou video in full screen, not in half screen
+  // know issue: can't identify sbs and ou video that don't need to change source size
+  if (videomode == RENDER_STEREO_MODE_SPLIT_VERTICAL)
+    sourceWidth = m_rs.Width() * 2;
+  if (videomode == RENDER_STEREO_MODE_SPLIT_HORIZONTAL)
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]:
```suggestion
  else if (videomode == RENDER_STEREO_MODE_SPLIT_HORIZONTAL)
```

videomode can be `RENDER_STEREO_MODE_SPLIT_VERTICAL`, `RENDER_STEREO_MODE_SPLIT_HORIZONTAL` (or something else), but not both at the same time.

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp [hunk 3]
```diff
@@ -16,6 +16,8 @@
 #include "cores/VideoPlayer/VideoRenderers/RenderManager.h"
 #include "windowing/GraphicContext.h"
 #include "guilib/GUIFontManager.h"
+#include "guilib/GUIComponent.h"
+#include "guilib/StereoscopicsManager.h"
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: this kind of globals hack is rather ugly. player is not supposed to depend on guilib.

Comment (taxigps) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: so how to get the 3d mode of the video file？ any advice?

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: It is passed with VideoPicture from VideoPlayerVideo to RenderManager

**Comment by taxigps** [comment]
@enen92 @Rechi @FernetMenta code updated

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp [hunk 4]
```diff
@@ -343,7 +348,19 @@ COverlay* CRenderer::Convert(CDVDOverlaySSA* o, double pts)
   int targetWidth = m_rv.Width();
   int targetHeight = m_rv.Height();
   int useMargin;
-
+  // Render subtitle of half-sbs and half-ou video in full screen, not in half screen
+  if (m_stereomode == "left_right" || m_stereomode == "right_left")
+  {
+    // only half-sbs video, sbs video don't need to change source size
+    if ((double)sourceWidth / sourceHeight < 1.2)
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: C++ style cast
```suggestion
    if (static_cast<double>(sourceWidth) / sourceHeight < 1.2)
```

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp [hunk 5]
```diff
@@ -343,7 +348,19 @@ COverlay* CRenderer::Convert(CDVDOverlaySSA* o, double pts)
...
+      sourceWidth = m_rs.Width() * 2;
+  }
+  else if (m_stereomode == "top_bottom" || m_stereomode == "bottom_top")
+  {
+    // only half-ou video, ou video don't need to change source size
+    if ((double)sourceWidth / sourceHeight > 2.5)
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]:
```suggestion
    if (static_cast<double>(sourceWidth) / sourceHeight > 2.5)
```

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp [hunk 6]
```diff
@@ -714,6 +714,7 @@ void CRenderManager::Render(bool clear, DWORD flags, DWORD alpha, bool gui)
     CRect src, dst, view;
     m_pRenderer->GetVideoRect(src, dst, view);
     m_overlays.SetVideoRect(src, dst, view);
+    m_overlays.SetStereoMode(m_stereomode);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp]: no need to call this on every cycle. move to Configure()

**Comment by taxigps** [comment]
@FernetMenta @Rechi update again

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp [hunk 7]
```diff
@@ -134,6 +134,7 @@ bool CRenderManager::Configure(const VideoPicture& picture, float fps, unsigned
     m_fps = fps;
     m_orientation = orientation;
     m_stereomode = picture.stereoMode;
+    m_overlays.SetStereoMode(m_stereomode);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp]: this is wrong Configure method. I meant the one without arguments. Thats why I explicitly wrote "Configure()"

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp [hunk 8]
```diff
@@ -331,6 +331,11 @@ void CRenderer::SetVideoRect(CRect &source, CRect &dest, CRect &view)
   m_rv = view;
 }
 
+void CRenderer::SetStereoMode(std::string &stereomode)
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp]: If passing by reference it should be always by const reference, if it is for consumption only.
```suggestion
void CRenderer::SetStereoMode(const std::string& stereomode)
```

**Comment by taxigps** [comment]
@FernetMenta updated. hope it's right this time.

**Comment by Rechi** [comment]
wrong revision was build
jenkins build this please

**Review by FernetMenta — APPROVED** [review]
looks good. thanks!

**Pull request merged**